### ISSUE TITLE
Support H265/HEVC for Livestreams

### ISSUE THREAD

**Issue by MattyBoombalatty** [Chocobozzz/PeerTube#6330]
### Describe the problem to be solved

After spending the good part of today editing a plugin I forked to enable H264 NVENC encoding, I realized that H265 only seems to be supported in VODs, not Livestreams. I'm not sure if this is an issue on my end, or with the browser (I tried several). The replay transcodes fine, and is viewable as normal

### Describe the solution you would like

Enable H265/HEVC live video playback in browser.

**Comment by Chocobozzz** [comment]
> I realized that H265 only seems to be supported in VODs, not Livestreams. 

Can you describe the problem? Is the player unable to play the live stream? Or is the server unable to transcode it?

**Comment by MattyBoombalatty** [comment]
I'm able to transcode it fine - I have a plugin and it works pretty well. However, the browser itself will not play the video during a livestream, only during the replay.

1. Livestream with HEVC profile
2. Video shows as live, but there is no video on screen
3. Terminate livestream 
4. Replay video, which plays back in browser correctly

The only issue is that the video does not play back in browser. Here's a link to my plugin if you wanted a better idea of how it works, in case I missed something.

https://gitea.nicecrew.digital/matty/peertube-plugin-nctv-nvenc-transcode

**Comment by MattyBoombalatty** [comment]
Here's the error - I was incorrect, this does happen with VODs too, on some browsers but I think it depends on what format they were uploaded in. Not sure.

```
Failed to play video
The video failed to play due to technical issues.

HLS.js error: mediaError - fatal: true - manifestIncompatibleCodecsError
```

**Comment by MattyBoombalatty** [comment]
Update:

H265 encoded VODs will play on Windows and Android mobile devices across Chromium based browsers (Brave, Chrome, Edge) Live videos will not. Perhaps it's the FFMPEG options that are passed in? Who knows.

H265 encoded VODs do *not* play on Linux browsers (tested Firefox, Chrome and Brave). Unfortunately I believe this is simply a limitation of supported devices. It would still be nice to have Live videos encoded in H265 play on supported browsers and hardware, so I'll leave this open in an attempt to assist in future proofing PeerTube.

**Comment by MattyBoombalatty** [comment]
There is no error. Audio plays, but not video. It's just a black screen.

**Comment by emansom** [comment]
I'd recommend switching to VP9 instead, broader compatibility and should be supported by NVENC as well.

**Comment by MattyBoombalatty** [comment]
We already do use H264 NVENC and it works fine. I was just hoping to be able to support hevc_nvenc for future proofing and bandwidth constraints.

**Comment by emansom** [comment]
> We already do use H264 NVENC and it works fine. I was just hoping to be able to support hevc_nvenc for future proofing and bandwidth constraints. 

HEVC/H265 is dead due to licensing issues. Use AV1 for future proofing and bandwidth saving. 😉

**Comment by emansom** [comment]
> Well unfortunately the GPU in my server cares less about licensing issues and more about encoder support, of which, AV1 is not supported. ;) 

It's an industry wide concern, this is why web browser compatibility of H265/HEVC isn't broad.

AV1 has a permissible license, allowing web browsers and media appliances to implement software fallbacks without fear of legal ramifications.

AV1 implements similar in goal features to H265/HEVC, which results in similar increase in quality and bandwidth saving.

If you have the available budget: the Intel Arc A310 is a very good choice as hardware accelerated encoder. It supports H264, VP9, H265/HEVC and AV1.

**Comment by MattyBoombalatty** [comment]
I am unfamiliar with Intel GPUs as they're relatively new to the market. I can budget for it, but that would also involve ensuring that there are Linux drives for it. I have a Tesla P4 at the moment and have been relatively pleased with it. I could, in theory, purchase one for research and see where it goes. I appreciate the heads up.

**Comment by emansom** [comment]
> I am unfamiliar with Intel GPUs as they're relatively new to the market. I can budget for it, but that would also involve ensuring that there are Linux drives for it. I have a Tesla P4 at the moment and have been relatively pleased with it. I could, in theory, purchase one for research and see where it goes. I appreciate the heads up. 

The kernel, Mesa drivers, VAAPI and ffmpeg support are all there in the latest stable releases. On rolling-release distro's like Fedora Rawhide/CentOS Stream 9 or Arch Linux it'd would work

There is the problem of fan control however, this is currently not yet implemented.

https://youtu.be/zjGieJci3xE

**Comment by MattyBoombalatty** [comment]
Much appreciated.

**Comment by Chocobozzz** [comment]
@MattyBoombalatty Can you provide a permanent live stream link + a vod video link so I can try to find what's the cause of your issue?

**Comment by MattyBoombalatty** [comment]
@Chocobozzz sure, when would be a good time for you? Did you want me to set up an HEVC encoded livestream and give you the link while it's going?

**Comment by Chocobozzz** [comment]
>  Did you want me to set up an HEVC encoded livestream and give you the link while it's going?

Yes please

**Comment by MattyBoombalatty** [comment]
Okay, I'll do that tonight. Our time differences make this more difficult :smile:

**Comment by Chocobozzz** [comment]
Yes

**Comment by Chocobozzz** [comment]
Thanks, unfortunately the issue seems more complex that I thought. Can you try AV1 and tell me if it works?

**Comment by MattyBoombalatty** [comment]
The transcoding GPU on my server does not support AV1, unfortunately. In the future I will likely upgrade GPUs, but I'll see if I can put together a plugin for it. Thanks for your help!

**Comment by SimplyCorbett** [comment]
I don’t know about live-streaming but HEVC VOD worked fine for me with the transcoding debug plugin. I think I had to modify a line or two in the source code as well.

Linux web browsers work but have to be compiled with hevc support. For windows you need to download the hevc codec in the windows store and use edge.

im not sure if this has changed since hevc has gained more mainline support however with the web browsers as this was over a year ago.

**Comment by Chocobozzz** [comment]
@MattyBoombalatty What's the status of this issue? It seems your VOD video works on my side (hls.js upgrades may have helped)

**Comment by MattyBoombalatty** [comment]
I will have to test. I haven't done anything else with HEVC since my last comment on this issue.

**Comment by MattyBoombalatty** [comment]
@Chocobozzz Yeah it looks like it works on basically everything except for Linux browsers, probably because HEVC is a proprietary codec. From what I've been reading, there are couple browsers with HEVC enabled but may require some extra stuff installed on the OS. iOS, Android and Windows desktop (including Brave browser) all seem to work fine with HEVC.

**Comment by Chocobozzz** [comment]
Thank you for your feedback, then I think we can close this issue :)

**Comment by beedell-roke** [comment]
> [`#issuecomment-3135986403`](https://github.com/Chocobozzz/PeerTube/issues/6330#issuecomment-3135986403)

@MattyBoombalatty, it's OS-specific, when the browser is installed from the distribution repositories. If installed from Flatpak, Snap, or a vendor-specific repository that provides the relevant libraries, it'll work. You'll experience an identical problem if you utilise certain editions of Windows, too (I believe, the N Edition, without the Media Pack).

**Comment by warnerbryce** [comment]
I have the same problem on Peertube 8.0.0
If i stream with OBS -> H.265 NVENC, the result on Peertube is Audio Channel only. 
I tried with Transcoding Runners and without same result. 

What kind of log you need ?
